python/katana/lonestar/analytics/assortativity.py

```diff
@@ -0,0 +1,106 @@
+from katana.atomic import atomic_add, GAccumulator
+from katana.loops import do_all, do_all_operator
+from katana.property_graph import PropertyGraph
+from katana.lonestar.analytics.calculate_degree import calculate_degree
+
+
+@do_all_operator()
+def sum_degree_operator(
+    graph: PropertyGraph,
+    source_degree,
+    sum_source: GAccumulator[np.uint64],
+    destination_degree,
+    sum_destination: GAccumulator[np.uint64],
+    nid,
+):
+    for edge in graph.edges(nid):
+        sum_source.update(source_degree[nid])
+        dst = graph.get_edge_dst(edge)
+        sum_destination.update(destination_degree[dst])
+
+
+def average_degree(graph: PropertyGraph, num_edges: int, source_degree, destination_degree):
+    """
+    Calculate the average in or out degree for the source and destination nodes
+    Returns the result as a tuple in the form (average degree for source, average degree for destination)
+    """
+    sum_source_degrees = GAccumulator[np.uint64](0)
+    sum_destination_degrees = GAccumulator[np.uint64](0)
+    do_all(
+        range(graph.num_nodes()),
+        sum_degree_operator(graph, source_degree, sum_source_degrees, destination_degree, sum_destination_degrees),
+        steal=True,
+    )
+    return (sum_source_degrees.reduce() / num_edges, sum_destination_degrees.reduce() / num_edges)
+
+
+@do_all_operator()
+def degree_assortativity_coefficient_operator(
+    graph: PropertyGraph,
+    source_degree,
+    source_average,
+    destination_degree,
+    destination_average,
+    product_of_dev: GAccumulator[float],
+    square_of_source_dev: GAccumulator[float],
+    square_of_destination_dev: GAccumulator[float],
+    nid,
+):
+    # deviation of source node from average
+    source_dev = source_degree[nid] - source_average
+    for edge in graph.edges(nid):
+        dst = graph.get_edge_dst(edge)
+        destination_dev = destination_degree[dst] - destination_average
+        product_of_dev.update(source_dev * destination_dev)
+        square_of_source_dev.update(source_dev * source_dev)
+        square_of_destination_dev.update(destination_dev * destination_dev)
+
+
+def degree_assortativity_coefficient(graph: PropertyGraph, x="out", y="in", weight=None):
+    """
+    Calculates and returns the degree assortativity of a given graph.
+    Paramaters:
+       * graph: the PropertyGraph to be analyzed
+       * x: description of degree type to consider for the source node on an edge
+            expected values are "in" or "out"
+       * y: description the degree type to consider for the destination node on an edge
+            expected values are "in" or "out"
+       * weight (optional): edge property to use if using weighted degrees
+    """
+    # get the tables associated with the degree types of the source and destination nodes
+    calculate_degree(graph, "temp_in", "temp_out", weight)
+    source_degree = graph.get_node_property("temp_" + x)
+    destination_degree = graph.get_node_property("temp_" + y)
```

Comment:
Using strings and then just appending them to produce some internal names is going to produce really weird error messages if someone uses the function wrong. Can you make and enum type for x and y? (https://docs.python.org/3/library/enum.html#creating-an-enum) That way errors will be much easier to catch in code that calls this. Also better names for x and y would be good. Maybe something like "source_degree_kind"? Up to you.

Reply:
I think I may have misunderstood what you mean. Should I create a separate class entirely for this enum, or can it be an inner class?